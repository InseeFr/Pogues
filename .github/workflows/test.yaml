name: Pogues Test CI

on:
  push:
    branches: ['main', 'next']
    paths-ignore: ['next/**']
  pull_request:
    types: [opened, synchronize, reopened]
    paths-ignore: ['next/**']

jobs:
  test:
    name: Test
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v4
      - uses: actions/setup-node@v4
        with:
          node-version: 20
      - run: yarn
      # Run tests sonar reports
      - run: yarn test:coverage
      - name: Upload coverage artifact
        uses: actions/upload-artifact@v4
        with:
          name: coverage
          path: coverage

  sonarcloud:
    name: SonarCloud
    runs-on: ubuntu-latest
    needs: test
    steps:
      - uses: actions/checkout@v4
        with:
          fetch-depth: 0
      - name: Download coverage artifact
        uses: actions/download-artifact@v4
        with:
          name: coverage
          path: coverage
      - name: SonarCloud Scan
        uses: SonarSource/sonarqube-scan-action@v4
        env:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
          SONAR_TOKEN: ${{ secrets.SONAR_TOKEN }}
