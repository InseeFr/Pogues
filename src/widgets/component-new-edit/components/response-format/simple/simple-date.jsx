import React from 'react';

import { connect } from 'react-redux';
import { Field, FormSection, formValueSelector } from 'redux-form';
import { date, required as funcRequired } from 'redux-form-validators';

import { DATATYPE_NAME } from '../../../../../constants/pogues-constants';
import GenericOption from '../../../../../forms/controls/generic-option';
import Input from '../../../../../forms/controls/input';
import Select from '../../../../../forms/controls/select';
import Dictionary from '../../../../../utils/dictionary/dictionary';

const { DATE } = DATATYPE_NAME;

const requiredValidationFunction = funcRequired({
  message: Dictionary.validationRequired,
});

function ResponseFormatDatatypeDate({
  format,
  type,
  isCollectedVariables,
  formatTableList,
  formatTable,
  formatCollectedVariables,
  name,
  required,
  readOnly,
}) {
  const formatInit =
    (type !== 'TABLE' && format) ||
    (type === 'TABLE' && isCollectedVariables
      ? formatCollectedVariables
      : formatTableList || formatTable);
  return (
    <FormSection name={name}>
      <div className="response-format-datatype-date">
        <Field
          name="format"
          id="date_format"
          component={Select}
          label={Dictionary.date_format}
          required={required}
          disabled={readOnly}
        >
          <GenericOption key="" value="">
            {Dictionary.dateinitial}
          </GenericOption>
          <GenericOption key="YYYY-MM-DD" value="YYYY-MM-DD">
            {Dictionary.dateddmmyyyy}
          </GenericOption>
          <GenericOption key="YYYY-MM" value="YYYY-MM">
            {Dictionary.datemmyyyy}
          </GenericOption>
          <GenericOption key="YYYY" value="YYYY">
            {Dictionary.dateyyyy}
          </GenericOption>
        </Field>
        <div hidden={formatInit === ''}>
          <Field
            name="minimum"
            type="text"
            step="any"
            component={Input}
            label={Dictionary.minimum}
            validate={[
              requiredValidationFunction,
              date({
                format: formatInit.toLowerCase(),
                message: Dictionary.formatDate ? Dictionary.formatDate : '',
              }),
            ]}
            disabled={readOnly}
            required={required}
          />
          <Field
            name="maximum"
            type="text"
            step="any"
            component={Input}
            label={Dictionary.maximum}
            validate={[
              requiredValidationFunction,
              date({
                format: formatInit.toLowerCase(),
                message: Dictionary.formatDate ? Dictionary.formatDate : '',
              }),
            ]}
            disabled={readOnly}
            required={required}
          />
        </div>
      </div>
    </FormSection>
  );
}

ResponseFormatDatatypeDate.defaultProps = {
  name: DATE,
  readOnly: false,
  required: true,
};

const mapStateToProps = (state) => {
  const selector = formValueSelector('component');
  return {
    formatCollectedVariables: selector(state, 'collectedVariables.DATE.format'),
    formatTable: selector(
      state,
      'responseFormat.TABLE.LIST_MEASURE.SIMPLE.DATE.format',
    ),
    formatTableList: selector(
      state,
      'responseFormat.TABLE.MEASURE.SIMPLE.DATE.format',
    ),
    format: selector(state, 'responseFormat.SIMPLE.DATE.format'),
    type: selector(state, 'responseFormat.type'),
  };
};
export { ResponseFormatDatatypeDate };

export default connect(mapStateToProps)(ResponseFormatDatatypeDate);
