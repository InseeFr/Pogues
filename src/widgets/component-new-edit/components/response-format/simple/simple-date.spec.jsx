import { render } from '@testing-library/react';
import { describe, expect, test } from 'vitest';

import { decoreFormField } from '../../../../../utils/test/test-utils';
import { ResponseFormatDatatypeDate } from './simple-date';

describe('Visualize Date Component "YYYY-MM-DD" format: ', () => {
  test('Should return the right HTML according for Date Field ', () => {
    const props = {
      componentId: 'component-id',
      format: 'YYYY-MM-DD',
    };
    const { container } = render(
      decoreFormField(<ResponseFormatDatatypeDate {...props} />),
    );
    expect(container).toMatchSnapshot();
  });
});

describe('Visualize Date Component "YYYY-MM" format: ', () => {
  test('Should return the right HTML according for Date Field ', () => {
    const props = {
      componentId: 'component-id',
      format: 'YYYY-MM',
    };
    const { container } = render(
      decoreFormField(<ResponseFormatDatatypeDate {...props} />),
    );
    expect(container).toMatchSnapshot();
  });
});

describe('Visualize Date Component "YYYY" format: ', () => {
  test('Should return the right HTML according for Date Field ', () => {
    const props = {
      componentId: 'component-id',
      format: 'YYYY',
    };
    const { container } = render(
      decoreFormField(<ResponseFormatDatatypeDate {...props} />),
    );
    expect(container).toMatchSnapshot();
  });
});
