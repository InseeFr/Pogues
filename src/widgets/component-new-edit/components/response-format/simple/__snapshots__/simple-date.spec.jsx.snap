// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Visualize Date Component "YYYY" format:  > Should return the right HTML according for Date Field  1`] = `
<div>
  <div
    class="response-format-datatype-date"
  >
    <div
      class="ctrl-select"
    >
      <label
        for="select-DATE.format"
      >
        Format
        <span
          class="ctrl-required"
        >
          *
        </span>
      </label>
      <div>
        <select
          id="select-DATE.format"
          name="DATE.format"
          placeholder="Format"
        >
          <option
            value=""
          >
            Date format
          </option>
          <option
            value="YYYY-MM-DD"
          >
            YYYY-MM-DD
          </option>
          <option
            value="YYYY-MM"
          >
            YYYY-MM
          </option>
          <option
            value="YYYY"
          >
            YYYY
          </option>
        </select>
      </div>
    </div>
    <div>
      <div
        class="ctrl-input "
      >
        <label
          for="input-DATE.minimum"
        >
          Minimum
          <span
            class="ctrl-required"
          >
            *
          </span>
        </label>
        <div>
          <input
            id="input-DATE.minimum"
            name="DATE.minimum"
            placeholder="Minimum"
            step="any"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        class="ctrl-input "
      >
        <label
          for="input-DATE.maximum"
        >
          Maximum
          <span
            class="ctrl-required"
          >
            *
          </span>
        </label>
        <div>
          <input
            id="input-DATE.maximum"
            name="DATE.maximum"
            placeholder="Maximum"
            step="any"
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Visualize Date Component "YYYY-MM" format:  > Should return the right HTML according for Date Field  1`] = `
<div>
  <div
    class="response-format-datatype-date"
  >
    <div
      class="ctrl-select"
    >
      <label
        for="select-DATE.format"
      >
        Format
        <span
          class="ctrl-required"
        >
          *
        </span>
      </label>
      <div>
        <select
          id="select-DATE.format"
          name="DATE.format"
          placeholder="Format"
        >
          <option
            value=""
          >
            Date format
          </option>
          <option
            value="YYYY-MM-DD"
          >
            YYYY-MM-DD
          </option>
          <option
            value="YYYY-MM"
          >
            YYYY-MM
          </option>
          <option
            value="YYYY"
          >
            YYYY
          </option>
        </select>
      </div>
    </div>
    <div>
      <div
        class="ctrl-input "
      >
        <label
          for="input-DATE.minimum"
        >
          Minimum
          <span
            class="ctrl-required"
          >
            *
          </span>
        </label>
        <div>
          <input
            id="input-DATE.minimum"
            name="DATE.minimum"
            placeholder="Minimum"
            step="any"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        class="ctrl-input "
      >
        <label
          for="input-DATE.maximum"
        >
          Maximum
          <span
            class="ctrl-required"
          >
            *
          </span>
        </label>
        <div>
          <input
            id="input-DATE.maximum"
            name="DATE.maximum"
            placeholder="Maximum"
            step="any"
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Visualize Date Component "YYYY-MM-DD" format:  > Should return the right HTML according for Date Field  1`] = `
<div>
  <div
    class="response-format-datatype-date"
  >
    <div
      class="ctrl-select"
    >
      <label
        for="select-DATE.format"
      >
        Format
        <span
          class="ctrl-required"
        >
          *
        </span>
      </label>
      <div>
        <select
          id="select-DATE.format"
          name="DATE.format"
          placeholder="Format"
        >
          <option
            value=""
          >
            Date format
          </option>
          <option
            value="YYYY-MM-DD"
          >
            YYYY-MM-DD
          </option>
          <option
            value="YYYY-MM"
          >
            YYYY-MM
          </option>
          <option
            value="YYYY"
          >
            YYYY
          </option>
        </select>
      </div>
    </div>
    <div>
      <div
        class="ctrl-input "
      >
        <label
          for="input-DATE.minimum"
        >
          Minimum
          <span
            class="ctrl-required"
          >
            *
          </span>
        </label>
        <div>
          <input
            id="input-DATE.minimum"
            name="DATE.minimum"
            placeholder="Minimum"
            step="any"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        class="ctrl-input "
      >
        <label
          for="input-DATE.maximum"
        >
          Maximum
          <span
            class="ctrl-required"
          >
            *
          </span>
        </label>
        <div>
          <input
            id="input-DATE.maximum"
            name="DATE.maximum"
            placeholder="Maximum"
            step="any"
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Visualize Date Component:  > Should return the right HTML according for Date Field  1`] = `
<div>
  <div
    class="response-format-datatype-date"
  >
    <div
      class="ctrl-select"
    >
      <label
        for="select-DATE.format"
      >
        Format
        <span
          class="ctrl-required"
        >
          *
        </span>
      </label>
      <div>
        <select
          id="select-DATE.format"
          name="DATE.format"
          placeholder="Format"
        >
          <option
            value=""
          >
            Date format
          </option>
          <option
            value="YYYY-MM-DD"
          >
            YYYY-MM-DD
          </option>
          <option
            value="YYYY-MM"
          >
            YYYY-MM
          </option>
          <option
            value="YYYY"
          >
            YYYY
          </option>
        </select>
      </div>
    </div>
    <div>
      <div
        class="ctrl-input "
      >
        <label
          for="input-DATE.minimum"
        >
          Minimum
          <span
            class="ctrl-required"
          >
            *
          </span>
        </label>
        <div>
          <input
            id="input-DATE.minimum"
            name="DATE.minimum"
            placeholder="Minimum"
            step="any"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        class="ctrl-input "
      >
        <label
          for="input-DATE.maximum"
        >
          Maximum
          <span
            class="ctrl-required"
          >
            *
          </span>
        </label>
        <div>
          <input
            id="input-DATE.maximum"
            name="DATE.maximum"
            placeholder="Maximum"
            step="any"
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
